services/beaconchain/src/main/java/tech/pegasys/artemis/services/beaconchain/BeaconChainController.java

```diff
@@ -281,21 +281,19 @@ private void onStoreInitializedEvent(final StoreInitializedEvent event) {
   @Subscribe
   @SuppressWarnings("unused")
   private void onTick(Date date) {
-    if (!testMode && !stateProcessor.isGenesisReady()) {
+    if (chainStorageClient.isPreGenesis()) {
       return;
     }
     final UnsignedLong currentTime = UnsignedLong.valueOf(date.getTime() / 1000);
-    if (chainStorageClient.getStore() != null) {
-      final Store.Transaction transaction = chainStorageClient.startStoreTransaction();
-      on_tick(transaction, currentTime);
-      transaction.commit().join();
-      final UnsignedLong nextSlotStartTime =
-          chainStorageClient
-              .getGenesisTime()
-              .plus(nodeSlot.times(UnsignedLong.valueOf(SECONDS_PER_SLOT)));
-      if (chainStorageClient.getStore().getTime().compareTo(nextSlotStartTime) >= 0) {
-        processSlot();
-      }
+    final Store.Transaction transaction = chainStorageClient.startStoreTransaction();
+    on_tick(transaction, currentTime);
+    transaction.commit().join();
+    final UnsignedLong nextSlotStartTime =
+        chainStorageClient
+            .getGenesisTime()
+            .plus(nodeSlot.times(UnsignedLong.valueOf(SECONDS_PER_SLOT)));
+    if (chainStorageClient.getStore().getTime().compareTo(nextSlotStartTime) >= 0) {
+      processSlot();
```

Comment:
One thing I noticed in this PR, unrelated to this PR is that process slot gets called only when storeTime is greater than nextSlotStartTime. Since, at least with the current constants, onTick() gets called twice every second, in the worst-case scenario we're losing 500 ms processing time for each slot. I think we should fix this. Here's the ticket: https://pegasys1.atlassian.net/secure/RapidBoard.jspa?rapidView=67&modal=detail&selectedIssue=BC-244
One idea is to have a SlotScheduler class that has its own timer which gets kicked off exactly at genesis. This class could post SlotEvents, BroadcastAttestationEvents and BroadcastAggregateEvents just like the processSlot method here, however, it would post those events way more precisely.

Reply:
Yep, definitely some improvements to be made around here.  I've been thinking of having a separate `TimeService` that fires time, slot and epoch events in.  Would need to do more thinking and playing to know if that will really work, but it's appealing in terms of being able to start doing time travel in tests because it's a single source of where time is up to.

Reply:
Agreed. A separate TimeService is a good direction.